### PR TITLE
ITransactionManager

### PR DESCRIPTION
**Pull request by zinkoviktor**

### CODE REVIEW

File: NetCoreRestApi/DataLayer.EF/EfTransactionManager.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿using DataLayer.interfaces;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Storage;
+
+namespace DataLayer.EF
+{
+    public class EfTransactionManager : ITransactionManager
+    {
+        private readonly DbContext _dbContext;
+        private IDbContextTransaction transaction1;
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer.EF/EfTransactionManager.cs]: Rename to _transactionManager

File: NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+using DataLayer.Models;
+
+namespace DataLayer.interfaces
+{
+    public interface ITransactionManager
+    {
+        void BeginTransaction();
```
Comment (mdemii) [inline NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs:7]: Add Dispose()

**Review by mdemii — CHANGES_REQUESTED** [review]
Add mock transaction manager

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs [hunk 3]
```diff
@@ -27,36 +28,51 @@ public virtual TModel GetById(TId id)
             return Сonverter.ConvertTo(entity);
         }
 
-        public virtual IQueryable<TModel> GetAll()
+        public virtual IQueryable<TModel> GetAll(FilterParameters filter)
         {
-            var entities = DbSet.ToList();
-            return Сonverter.ConvertTo(entities).AsQueryable();
+            var entities = (filter == null) ? 
+                DbSet.ToList().AsQueryable() : DbSet.Skip((filter.PageIndex - 1) * filter.PageSize)
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: Remove  .ToList().AsQueryable()

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs [hunk 4]
```diff
@@ -27,36 +28,51 @@ public virtual TModel GetById(TId id)
...
+                                                                         .Take(filter.PageSize);
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: Move take into one line with skip

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs [hunk 5]
```diff
@@ -27,36 +28,51 @@ public virtual TModel GetById(TId id)
...
+            var models = Сonverter.ConvertTo(entities);
+            return models.AsQueryable();
         }
 
         public virtual IQueryable<TModel> Create(IEnumerable<TModel> models)
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: Replace IQueryable with IEnumerable

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs [hunk 6]
```diff
@@ -2,19 +2,53 @@
...
-            _categoryRepository = categoryRepository;
+            var createdModels = new List<ProductModel>();
+
+            foreach (var model in models)
+            {
+                var foundProduct = DbSet.FirstOrDefault(p => p.Name == model.Name);
+
+                if (foundProduct == null)
+                {
+                    var product = Сonverter.ConvertFrom(model);
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: rename product into productEntity

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs [hunk 7]
```diff
@@ -2,19 +2,53 @@
...
+                    {
+                        var foundCategory = _categoryDbSet.FirstOrDefault(
+                            c => c.Name == productCategory.Category.Name);
+
+                        if (foundCategory != null)
+                        {
+                            productCategory.CategoryId = foundCategory.Id;
+                            productCategory.Category = foundCategory;
+                        }
+                    }
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: Refactor method into UpdateProductCategories

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs [hunk 8]
```diff
@@ -2,19 +2,53 @@
...
-                : base(dbContext, converter)
+        private DbSet<CategoryEntity> _categoryDbSet;
+
+        public ProductRepository(IDbContext dbContext, IConverter<ProductEntity, ProductModel> converter)
+            : base(dbContext, converter)
+        {
+            _categoryDbSet = dbContext.GetDbSet<CategoryEntity>();
+        }
+
+        public override IQueryable<ProductModel> Create(IEnumerable<ProductModel> models)
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: IEnumerable

File: NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs (hunk 2 above)
Comment (mdemiy) [reply NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs:7]: ?

File: NetCoreRestApi/DataLayer/Repositories/Interfaces/IRepository.cs [hunk 9]
```diff
@@ -8,7 +9,7 @@ public interface IRepository<TModel, TId>
         where TModel : BaseModel<TId>
     {
         TModel GetById(TId id);
-        IQueryable<TModel> GetAll();
+        IQueryable<TModel> GetAll(FilterParameters filter);
         IQueryable<TModel> Create(IEnumerable<TModel> models);
         IQueryable<TModel> Update(IEnumerable<TModel> models);
         IQueryable<TModel> Delete(IEnumerable<TModel> models);
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer/Repositories/Interfaces/IRepository.cs]: IEnumerable

File: NetCoreRestApi/DataLayer/UnitOfWorks/BaseUnitOfWork.cs [hunk 10]
```diff
@@ -1,51 +1,66 @@
...
+        public virtual IQueryable<TModel> GetAll(FilterParameters filter)
         {
-            return _repository.GetAll();
+            return Repository.GetAll(filter);
         }
 
         public virtual IEnumerable<TModel> Create(IEnumerable<TModel> models)
         {
-            return _repository.Create(models);
+            TransactionManager.BeginTransaction();
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer/UnitOfWorks/BaseUnitOfWork.cs:34]: Add logic for checking if BeginTransaction has been started

File: NetCoreRestApi/DataLayer/UnitOfWorks/FilterParameters.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿namespace DataLayer.UnitOfWorks
+{
+    public class FilterParameters
+    {
+        public int PageIndex { get; set; }
```
Comment (mdemiy) [inline NetCoreRestApi/DataLayer/UnitOfWorks/FilterParameters.cs]: Rename into PageNumber

File: NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs [hunk 12]
```diff
@@ -16,13 +18,50 @@ public class CategoriesController : GenericController<CategoryDto, CategoryModel
...
             var categoriesDTO = Converter.ConvertFrom(categoryModels.ToList());
 
             return Ok(categoriesDTO);
         }
+
+        [HttpGet]
+        public IActionResult Get()
+        {
+            return Get(null);
+        }
```
Comment (mdemiy) [inline NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs]: Remove this you have one with filter

File: NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs [hunk 13]
```diff
@@ -16,13 +18,50 @@ public class CategoriesController : GenericController<CategoryDto, CategoryModel
         {
         }
 
-        [HttpGet]
-        public IActionResult Get()
+        [HttpGet("/")]
```
Comment (mdemiy) [inline NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs]: Remove "/"

File: NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs [hunk 14]
```diff
@@ -16,13 +18,50 @@ public class CategoriesController : GenericController<CategoryDto, CategoryModel
...
+        {
+            return Get(null);
+        }
+
+        [HttpPost]
+        public IActionResult Create([FromBody] IEnumerable<CategoryDto> categories)
+        {
+            var categoryModels = Converter.ConvertTo(categories);
+            var createdModels = Manager.Create(categoryModels);
+            Manager.Save();
```
Comment (mdemiy) [inline NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs]: Redundant

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs [hunk 15]
```diff
@@ -16,13 +18,50 @@ public class ProductsController : GenericController<ProductDto, ProductModel>
...
         public IActionResult Get()
         {
-            var productModels = Manager.GetAll();
-            var productsDTO = Converter.ConvertFrom(productModels.ToList());
+            return Get(null);
+        }
 
-            return Ok(productsDTO);
+        [HttpPost]
+        public IActionResult Create([FromBody] IEnumerable<ProductDto> products)
```
Comment (mdemiy) [inline NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: rename products into productDtos

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs [hunk 16]
```diff
@@ -16,13 +18,50 @@ public class ProductsController : GenericController<ProductDto, ProductModel>
         {
         }
 
+        [HttpGet("/")]
+        public IActionResult Get([FromQuery] FilterParameters filter)
+        {            
+            var productModels = Manager.GetAll(filter);
+            var productsDTO = Converter.ConvertFrom(productModels.ToList());
```
Comment (mdemiy) [inline NetCoreRestApi/WebAPI/Controllers/ProductsController.cs:25]: Do innevstigation to use ProjectTo to return IQueryable

**Review by mdemiy — CHANGES_REQUESTED** [review]
Fix commments

File: NetCoreRestApi/DataLayer.EF/EfTransactionManager.cs (hunk 1 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/EfTransactionManager.cs]: done

File: NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs (hunk 2 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer/Interfaces/ITransactionManager.cs:7]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs (hunk 3 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs (hunk 4 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs (hunk 5 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/BaseRepository.cs]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs (hunk 6 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs (hunk 7 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: done

File: NetCoreRestApi/DataLayer/Repositories/Interfaces/IRepository.cs (hunk 9 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer/Repositories/Interfaces/IRepository.cs]: done

File: NetCoreRestApi/DataLayer/UnitOfWorks/BaseUnitOfWork.cs (hunk 10 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer/UnitOfWorks/BaseUnitOfWork.cs:34]: done

File: NetCoreRestApi/DataLayer/UnitOfWorks/FilterParameters.cs (hunk 11 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer/UnitOfWorks/FilterParameters.cs]: done

File: NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs (hunk 8 above)
Comment (zinkoviktor) [reply NetCoreRestApi/DataLayer.EF/Repositories/ProductRepository.cs]: done

File: NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs (hunk 12 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Controllers/CategoriesController.cs]: done

File: NetCoreRestApi/WebAPI/Controllers/ProductsController.cs (hunk 15 above)
Comment (zinkoviktor) [reply NetCoreRestApi/WebAPI/Controllers/ProductsController.cs]: done

**Review by mdemiy — APPROVED** [review]

**Review by mdemiy — APPROVED** [review]

**Pull request merged**